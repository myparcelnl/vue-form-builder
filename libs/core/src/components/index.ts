export {default as MagicForm} from './MagicForm.vue';
