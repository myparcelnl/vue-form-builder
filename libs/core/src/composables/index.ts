export * from './useElementHooks';
export * from './useFormBuilder';
export * from './useLifecycleHooks';
export * from './useTestAttributes';
