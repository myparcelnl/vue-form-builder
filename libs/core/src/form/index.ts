export * from './Form';
export * from './Form.types';
export * from './defaultFormConfiguration';
export * from './defineField';
export * from './defineForm';
export * from './interactive-element';
export * from './plain-element';
export * from './validator';
