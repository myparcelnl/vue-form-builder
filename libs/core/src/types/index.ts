export * from './common.types';
export * from './component.types';
export * from './deprecated.types';
export * from './element.types';
export * from './form-builder.types';
export * from './form.types';
export * from './interactive-element.types';
export * from './plain-element.types';
export * from './validator.types';
