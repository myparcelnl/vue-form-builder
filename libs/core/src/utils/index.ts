export * from './markComponentAsRaw';
export * from './useDynamicWatcher';
