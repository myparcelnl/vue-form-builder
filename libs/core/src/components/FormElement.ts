import {Component, PropType, Teleport, defineComponent, h} from 'vue';
import {FormInstance, InteractiveElementInstance} from '../form';
import {AnyElementInstance} from '../types';
import InteractiveElement from './InteractiveElement.vue';
import PlainElement from './PlainElement.vue';
import {isOfType} from '@myparcel/ts-utils';
import {useTestAttributes} from '../composables/useTestAttributes';

// noinspection JSUnusedGlobalSymbols
export default defineComponent({
  name: 'FormElement',
  props: {
    form: {
      type: Object as PropType<FormInstance>,
      required: true,
    },
    element: {
      type: Object as PropType<AnyElementInstance>,
      required: true,
    },
  },

  render() {
    let component: Component = h(
      isOfType<InteractiveElementInstance>(this.element, 'ref') ? InteractiveElement : PlainElement,
      {
        ...useTestAttributes(this.element),
        element: this.element,
      },
    );

    const hasOwnWrapper = typeof this.element.wrapper !== 'boolean';
    const inheritsWrapper = this.element.wrapper === true && this.form.config.field?.wrapper;

    if (inheritsWrapper || hasOwnWrapper) {
      const childComponent = component;

      component = h(
        (hasOwnWrapper ? this.element.wrapper : this.form.config.field?.wrapper) as Component,
        {
          ...this.$attrs,
          ...this.form.config.fieldDefaults.attributes,
          element: this.element,
        },
        () => [childComponent],
      );
    }

    if (this.element.teleportSelector) {
      // @ts-expect-error Teleport has an extra property which doesn't fit in the Component type
      return h(Teleport, {to: this.element.teleportSelector}, [component]);
    }

    return component;
  },
});
