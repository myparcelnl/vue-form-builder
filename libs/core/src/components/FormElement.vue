<template>
  <component
    :is="element.component"
    v-model="model"
    v-bind="attributes">
    <template
      v-for="name in Object.keys($slots)"
      :key="name"
      #[name]="scope">
      <slot
        :name="name"
        v-bind="scope || {}" />
    </template>
  </component>
</template>

<script lang="ts" setup>
import {computed, toRefs, type Ref, unref} from 'vue';
import {type AnyElementInstance, type InteractiveElementInstance, type ComponentProps} from '../types';
import {createElementHooks} from '../composables';

const props = defineProps<{element: AnyElementInstance}>();

const propRefs = toRefs(props);

createElementHooks(propRefs.element.value);

/**
 * Collect attributes. Always adds `element.attributes`, but only adds `element` if element is a Vue component.
 */
const attributes = computed(() => {
  const elementProp = props.element.form?.config.field?.elementProp;

  const newProps: ComponentProps = {
    ...props.element.attributes,
  };

  if (typeof props.element.component !== 'string' && elementProp !== false) {
    newProps.element = props.element;
  }

  return newProps;
});

const model = computed({
  get() {
    return unref((props.element as InteractiveElementInstance).ref);
  },

  set(value) {
    (props.element as InteractiveElementInstance).ref = value as Ref;
  },
});
</script>
