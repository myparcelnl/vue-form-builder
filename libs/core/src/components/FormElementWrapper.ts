import {Component, PropType, Teleport, defineComponent, h} from 'vue';
import {AnyElementInstance} from '../types';
import FormElement from './FormElement.vue';
import {FormInstance} from '../form';
import {useTestAttributes} from '../composables';

// noinspection JSUnusedGlobalSymbols
export default defineComponent({
  name: 'FormElementWrapper',
  props: {
    form: {
      type: Object as PropType<FormInstance>,
      required: true,
    },
    element: {
      type: Object as PropType<AnyElementInstance>,
      required: true,
    },
  },

  render() {
    let component: Component = h(FormElement, {
      ...useTestAttributes(this.element),
      element: this.element,
    });

    const hasOwnWrapper = typeof this.element.wrapper !== 'boolean';
    const inheritsWrapper = this.element.wrapper === true && this.form.config.field?.wrapper;

    if (inheritsWrapper || hasOwnWrapper) {
      const childComponent = component;

      component = h(
        (hasOwnWrapper ? this.element.wrapper : this.form.config.field?.wrapper) as Component,
        {
          ...this.$attrs,
          ...this.form.config.fieldDefaults.attributes,
          element: this.element,
        },
        () => [childComponent],
      );
    }

    if (this.element.teleportSelector) {
      // @ts-expect-error Teleport has an extra property which doesn't fit in the Component type
      return h(Teleport, {to: this.element.teleportSelector}, [component]);
    }

    return component;
  },
});
