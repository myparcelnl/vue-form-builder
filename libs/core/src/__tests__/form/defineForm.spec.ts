import {afterEach, describe, expect, it, vi} from 'vitest';
import {flushPromises, mount} from '@vue/test-utils';
import {h, ref} from 'vue';
import {AnyElementConfiguration} from '../../types';
import {DEFAULT_FORM_CONFIGURATION} from '../../form';
import {MagicForm} from '../../components';
import TextInput from '../elements/TextInput.vue';
import {generateForm} from '../utils/generateForm';
import {useFormBuilder} from '../../composables';

describe('defining a form', () => {
  it('renders html elements', async () => {
    const form = generateForm({
      fields: [
        {
          name: 'named',
          component: 'input',
        },
        {
          component: 'br',
        },
        {
          name: 'text',
          component: TextInput,
          ref: ref('initial'),
        },
      ],
    });

    const wrapper = mount(MagicForm, {props: {form}});
    await flushPromises();
    const formElement = wrapper.find('form');
    expect(formElement.exists()).toBe(true);
    expect(formElement.attributes('id')).toBe('form');
    expect(formElement.find('input').exists()).toBe(true);
    expect(formElement.find('br').exists()).toBe(true);
    expect(formElement.find('input[name="text"]').exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('overriding defaults', () => {
    const formBuilder = useFormBuilder();

    const commonFields: AnyElementConfiguration[] = [
      {
        name: 'field',
        component: 'input',
        ref: ref(''),
        label: 'my_label_1',
      },
      {
        name: 'named',
        component: 'input',
        label: 'my_label_2',
      },
      {
        component: 'br',
      },
    ];

    afterEach(() => {
      formBuilder.defaults.value = DEFAULT_FORM_CONFIGURATION;
    });

    it('adds class to form', () => {
      formBuilder.defaults.value.form.attributes.class = 'form-class';

      const form = generateForm(commonFields);
      const wrapper = mount(MagicForm, {props: {form}});
      const formElement = wrapper.find('form');
      expect(formElement.classes()).toContain('form-class');
    });

    it('changes tag of outer form wrapper', () => {
      formBuilder.defaults.value.form.tag = 'div';
      formBuilder.defaults.value.form.attributes.class = 'form';

      const form = generateForm(commonFields);
      const wrapper = mount(MagicForm, {props: {form}});
      const formElement = wrapper.find('.form');

      expect(formElement.exists()).toBe(true);
      expect(formElement.element.tagName).toBe('DIV');
      expect(wrapper.find('form').exists()).toBe(false);
    });

    it('wraps fields with a wrapper element', () => {
      formBuilder.defaults.value.form.attributes.class = 'my-form';
      formBuilder.defaults.value.form.wrapper = h('div', {class: 'inner-wrapper'});

      const form = generateForm(commonFields);
      const wrapper = mount(MagicForm, {props: {form}});
      const wrapperElement = wrapper.find('.my-form > .inner-wrapper');

      expect(wrapper.html()).toMatchSnapshot();

      expect(wrapperElement.exists()).toBe(true);
      expect(wrapperElement.element.tagName).toBe('DIV');
    });

    it('adds attributes to each field', () => {
      formBuilder.defaults.value.fieldDefaults.attributes.class = 'default-field-class';
      formBuilder.defaults.value.fieldDefaults.attributes['aria-label'] = 'test';

      const form = generateForm(commonFields);
      const wrapper = mount(MagicForm, {props: {form}});

      const formElement = wrapper.find('form');

      expect(formElement.findAll('.default-field-class')).toHaveLength(3);
      expect(formElement.findAll('[aria-label="test"]')).toHaveLength(3);
    });

    it('makes fields lazy by default', () => {
      formBuilder.defaults.value.fieldDefaults.lazy = true;

      const form = generateForm(commonFields);
      expect(form.model.field.lazy).toBe(true);
    });

    it('makes fields optional by default', () => {
      formBuilder.defaults.value.fieldDefaults.optional = true;

      const form = generateForm(commonFields);
      expect(form.model.field.isOptional.value).toBe(true);
    });

    it('calls renderLabel if defined when rendering a label', () => {
      expect.assertions(1);
      const renderLabel = vi.fn((value: string): string => value.toUpperCase());

      formBuilder.defaults.value.renderLabel = renderLabel;

      const form = generateForm(commonFields);
      mount(MagicForm, {props: {form}});

      expect(renderLabel).toHaveBeenCalledTimes(1);
    });
  });
});
