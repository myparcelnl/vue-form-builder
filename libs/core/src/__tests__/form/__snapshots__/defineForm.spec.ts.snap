// Vitest Snapshot v1

exports[`defining a form > overriding defaults > wraps fields with a wrapper element 1`] = `
"<div id=\\"form\\" class=\\"my-form valid\\">
  <div class=\\"inner-wrapper\\"><input id=\\"field\\" modelvalue=\\"\\" label=\\"my_label_1\\" name=\\"field\\" errors=\\"\\" valid=\\"true\\" suspended=\\"false\\" data-test-id=\\"field\\" data-test-type=\\"interactive\\"><input id=\\"named\\" name=\\"named\\" label=\\"my_label_2\\" errors=\\"\\" data-test-id=\\"named\\" data-test-type=\\"plain\\"><br errors=\\"\\" data-test-type=\\"plain\\">
    <!--v-if-->
  </div>
</div>"
`;

exports[`defining a form > renders html elements 1`] = `
"<form id=\\"form\\" class=\\"valid\\"><input id=\\"named\\" name=\\"named\\" errors=\\"\\" data-test-id=\\"named\\" data-test-type=\\"plain\\"><br errors=\\"\\" data-test-type=\\"plain\\"><input id=\\"text\\" type=\\"text\\" name=\\"text\\" class=\\"\\">
  <!--v-if-->
  <!--v-if-->
</form>"
`;
